### DISCUSSION TITLE
Move and scale problems

### DISCUSSION THREAD

**Discussion by diegobenna** [didi/LogicFlow#1219] [Q&A]
Dear community, I write in English because I'm Italian. I hope you can help me.
I'm writing a program to let my students draw a flowchart with the idea of ​​transforming it into C++.
I'm using LogicFlow and I'm having problems. I hope you can help me.

1) I defined an edge like this:

```
var left_branch = {
        type:"EdgeLeft",
        sourceNodeId: new_node_condition.id,
        targetNodeId: output.id,
        sourceAnchorId: new_node_condition.id+"_left",
        targetAnchorId: output.id+"_left",
        text: { x: x_condition-80, y: y_condition-10, value: "true" },
        startPoint: calculateConditionAnchorLeft(new_node_condition),
        endPoint: calculateEndConditionStillLeft(output),
        pointsList: [
            computeConditionAnchorLeft(new_node_condition),
            { x: x_condition - width_branch_if, y: y_condition },
            { x: x_condition - width_branch_if, y: y_condition + distance_nodes },
            computeEndConditionLeft Again(exit)
        ]
    };
```

So I use the pointsList. But when I use the lf.graphModel.moveNode function the edge loses its shape and is transformed. I use the lf.graphModel.moveNode function only to move down, y coordinate + 100 for both source and target nodes. So I expect the edge to remain of the previous form: only it would be shifted down by 100. I also see the text changing position with respect to the initial configurations.
What should I do to get this result?

2) I have a second problem: how can I retrieve the size of a node? How can I change its size? I would like to be able to dynamically push the content into the node and have the node scale to fit the content. Consequently I will have to change the shape of the whole flowchart according to the new width of the node.

Thank you very much if you could help me

**Comment by wumail** [comment]
> Dear community, I write in English because I'm Italian. I hope you can help me. I'm writing a program to let my students draw a flowchart with the idea of ​​transforming it into C++. I'm using LogicFlow and I'm having problems. I hope you can help me.
> 
> 1. I defined an edge like this:
> 
> ```
> var left_branch = {
>         type:"EdgeLeft",
>         sourceNodeId: new_node_condition.id,
>         targetNodeId: output.id,
>         sourceAnchorId: new_node_condition.id+"_left",
>         targetAnchorId: output.id+"_left",
>         text: { x: x_condition-80, y: y_condition-10, value: "true" },
>         startPoint: calculateConditionAnchorLeft(new_node_condition),
>         endPoint: calculateEndConditionStillLeft(output),
>         pointsList: [
>             computeConditionAnchorLeft(new_node_condition),
>             { x: x_condition - width_branch_if, y: y_condition },
>             { x: x_condition - width_branch_if, y: y_condition + distance_nodes },
>             computeEndConditionLeft Again(exit)
>         ]
>     };
> ```
> 
> So I use the pointsList. But when I use the lf.graphModel.moveNode function the edge loses its shape and is transformed. I use the lf.graphModel.moveNode function only to move down, y coordinate + 100 for both source and target nodes. So I expect the edge to remain of the previous form: only it would be shifted down by 100. I also see the text changing position with respect to the initial configurations. What should I do to get this result?
> 
> 2. I have a second problem: how can I retrieve the size of a node? How can I change its size? I would like to be able to dynamically push the content into the node and have the node scale to fit the content. Consequently I will have to change the shape of the whole flowchart according to the new width of the node.
> 
> Thank you very much if you could help me

hey diegobenna, 
for problem one, I don't actually know what's going on by your description, could you post your code about the way you use the moveNode function? 
And for problem two, i think you can directly access the `width` and `height` properties of the node model, you need to obtain the corresponding node model by calling `const nodeModel = lf.getNodeModelById(nodeId)`, then, you can access the width and height properties using `nodeModel.width` and `nodeModel.height` respectively. Additionally, you can modify the shape of the node by assigning new values to `nodeModel.width` and `nodeModel.height`, both width and height properties of nodeModel are observable, allowing for changes to be tracked (docs: https://docs.logic-flow.cn/docs/#/en/api/nodeModelApi).

**Reply by diegobenna** [reply]
Thanks to the availability. Here is the example. This is the initial flowchart.
![state1](https://github.com/didi/LogicFlow/assets/5632468/27d50f19-e96d-40c1-8299-a882ba81628e)

If I use the function to lower (bottom + 100) the nodes by 100px, the edges are transformed. I actually added one node at the top of the flow and lowered the others.
![state2](https://github.com/didi/LogicFlow/assets/5632468/b73c6495-5cf8-469c-8673-8e72d8a05160)

Basically the original pointList is transformed or lost.

**Reply by wumail** [reply]
I see, that is indeed a problem in LogicFlow where the pointList of an edge will automatically update when the source node or target node is transformed. The way the points change can be a little strange, and it would be difficult to apply your own rules to update the pointList of an edge if you don't modify the source code. This issue will be fixed in the near future.

**Reply by diegobenna** [reply]
But would it at least be possible to edit the poitslist manually? I tried but it doesn't work. If it is not possible then we will not be able to use logicflow in our business application. There is another problem: if I decide to modify the text content of a node, is it possible to modify the width so that the text is inside it? Following this, can I check if the node collides with others? Thank you

**Reply by wumail** [reply]
Sure, you can modify the pointList on the edge model manually. And if you know the width and height of the text content, you can directly modify the width and height on the node model. I think it's not difficult to track the width and height information of the text, and then trigger the modify action.

**Reply by diegobenna** [reply]
Thanks, but how do I change the pointsList dynamically and at runtime? i tried and failed.

How do I recover the width and height of the text contained in the node? From the documentation I could not understand. Thanks again for your availability.

Can you give me an example of the trigger youre referring to? Thank you

Regarding a third problem, I ask you for help: if I increase the width of a node and I have nodes next to them, they overlap. Is there any way to adjust the position of the nodes automatically so that no collisions happen?

**Reply by wumail** [reply]
1. You can register the relevant event and emit it at runtime. This allows you to update the pointList dynamically. 👀(https://docs.logic-flow.cn/docs/#/en/guide/basic/event, https://docs.logic-flow.cn/docs/#/en/api/eventCenterApi)
2.
```javascript
// When you update the text, it will trigger the 'text:update' event. In that event, you can obtain the target node and update its width and height. You can find more information about events in the documentation

const nodeModel = lf.getNodeModelById(nodeId)

function getActualWidthOfChars(text, options = {}) {
  const { size = 14, family = "Microsoft YaHei" } = options;
  const canvas = document.createElement("canvas");
  const ctx = canvas.getContext("2d");
  ctx.font = `${size}px ${family}`;
  const metrics = ctx.measureText(text);
  const actual = Math.abs(metrics.actualBoundingBoxLeft) + Math.abs(metrics.actualBoundingBoxRight);
  return Math.max(metrics.width, actual);
}

const text = 'TEST TEXT'
nodeModel.width = getActualWidthOfChars(text)
// To modify the height, you need to know the maximum length of the line. Then, you can update the nodeModel.height = (text.length / maxLengthOfLine) * lineHeight
``` 
3. It will be a little difficult because you will need to manually update the positions of all nodes around it to avoid collisions. You can simplify the finding process by using R-tree, and then calculate and update the positions of all these nodes as you want.

**Reply by diegobenna** [reply]
I have 3 files:

**App.js**
```
import LogicFlow from "@logicflow/core";
import "@logicflow/core/dist/style/index.css";

import { useEffect, useRef } from "react";
import MyNode from "./MyNode.js";
import MyEdge from "./MyEdge.js";

export default function App() {
  const refContainer = useRef();
  useEffect(() => {
    const logicflow = new LogicFlow({
      container: refContainer.current,
      grid: true,
      width: 1000,
      height: 500,
      //isSilentMode: true,
      //nodeTextEdit: false
    });
    logicflow.register(MyNode);  
    logicflow.register(MyEdge);  
  
    const nodes = {
      nodes: [
        {
          type: "MyNode",
          x: 0,
          y: 0,
          id: "nodo1",
          text: "Start",
        },
        {
          type: "MyNode",
          x: 200,
          y: 200,
          id: "nodo2",
          text: "End",
        },     
      ],
      edges: [
        {
          id:"nodo3",
          sourceNodeId: "nodo1",
          targetNodeId: "nodo2",
          text: { value: "text"},
          type: "MyEdge",          
          startPoint: { x: 0, y: 25 }, 
          endPoint: { x: 200, y: 175  }, 
          sourceAnchorId: "nodo1_down",
          targetAnchorId: "nodo2_up",                     
          pointsList: [
            { x: 0, y: 25 },
            { x: 30, y: 100  },
            { x: 100, y: 100  },
            { x: 200, y: 175  }
          ]           
        },
      ],
    };



// Aggiungi il collegamento dal nuovo nodo al nodo B
   

    logicflow.render(
      nodes
    );

    logicflow.focusOn({
      id: "nodo1",
    });       

    logicflow.on("edge:click", (data) => {
      console.log(data); 
      logicflow.graphModel.moveNode("nodo1", 0, -100, false);
      logicflow.graphModel.moveNode("nodo2", 0, -100, false);

    });  
    
    logicflow.on("node:click", (data) => {
      //console.log(logicflow.getModelById("nodo1"))
    });   
  }, []);
  
  return (
      <div ref={refContainer}></div>
  );}
```

**MyEdge.js**
```
import { PolylineEdge, PolylineEdgeModel } from "@logicflow/core";


class MyEdgeView extends PolylineEdge {}

class MyEdgeModel extends PolylineEdgeModel {

  initEdgeData(data) {
    super.initEdgeData(data);
    //this.visible=false;
    //this.draggable=false;
    //this.isHitable=false;
  }  

      getData() {
        const data = super.getData();
        data.sourceAnchorId = this.sourceAnchorId;
        data.targetAnchorId = this.targetAnchorId;   
        return data;
      }
}

export default {
  type: "MyEdge",
  view: MyEdgeView,
  model: MyEdgeModel,
};

```

**MyNode.js**
```
import { RectNode, RectNodeModel } from "@logicflow/core";

class MyNodeModel extends RectNodeModel { 

  initNodeData(data) {
    super.initNodeData(data);
    this.width = 50;
    this.height = 50;
  }

  getDefaultAnchor() {
    const { height, x, y, id } = this;
    return [
      {
        x,
        y: height/2 + y,
        name: 'down',
        id: `${id}_down`,
        edgeAddable: true
      },
      {
        x,
        y: y - height/2,
        name: 'up',
        id: `${id}_up`,
        edgeAddable: true
      }      
    ]
  } 
}

class MyNodeView extends RectNode {
 
}

export default {
  type: "MyNode",
  view: MyNodeView,
  model: MyNodeModel,
};
```
I'm missing something in understanding the library. I can't figure out how I can modify the list of edge points when I move nodes up. This is an example created for you who have been very kind up to now. Thank you

I try..
```
    logicflow.on("edge:click", (data) => {
      console.log(data); 
      logicflow.graphModel.moveNode("nodo1", 0, -100, false);
      logicflow.graphModel.moveNode("nodo2", 0, -100, false);
      console.log(logicflow.getEdgeDataById("nodo3"));
      logicflow.getEdgeDataById("nodo3").pointsList = [
        { x: 0, y: 25 - 100 },
        { x: 30, y: 100 - 100  },
        { x: 100, y: 100 - 100  },
        { x: 200, y: 175 - 100  }
      ]   
    }); 
```
    
but is not correct

**Reply by wumail** [reply]
Sorry, it's my mistake. The current approach won't work because any changes made to pointsList will trigger a re-render, and the edge's pointsList will be automatically updated based on the new source and target. If you can ensure that you always move two nodes together and keep them in the same relative position, you could override the initEdge method in the Edge Model, you can always adjust the pointsList by consistently subtracting or adding the same distance. However, this isn't really the desired outcome. Therefore, the conclusion is that we can't update pointsList manually at the moment. 😞

**Reply by diegobenna** [reply]
I am willing to manually calculate the poitslist coordinates. If so, would it be possible? Can you give me a practical example with code example? Could you possibly modify the library please?

**Reply by wumail** [reply]
Sorry, I'm busy these days. If you have a QQ number or WeChat, you can join the group chat, and you can ask questions to group members, they will be willing to solve your problems.

**Reply by diegobenna** [reply]
Thanks, i create now wechat id: wxid_a8wi332r67ya22

WQ for me is impossible

**Reply by wumail** [reply]
I can‘t provide a QR code since the number of members in the group has already exceeded 200. You can add the official WeChat account 'logic-flow,' and he will invite you in.

**Reply by diegobenna** [reply]
HI. Initially on WeChat everyone wrote and replied to me. Now the chat seems to stop. I don't see any more replies and messages. Is that so? Or is there some problem on my wechat?

**Reply by wumail** [reply]
Nah, that's not your problem, I saw the dialogue in the group chat, and I think most of them are at rest or taking a break on weekends. btw, we will optimize this problem before this Wednesday to provide a way for developers to change the way pointsList updates.
